syntax = "proto3";

package project;

import "google/protobuf/empty.proto";

message Project {
  string id = 1;
  string title = 2;
  string summary = 3;
  int32 priority = 4;
  string status = 5;
}

message CreateProjectRequest {
  string title = 1;
  string summary = 2;
  int32 priority = 3;
}

message UpdateProjectRequest {
  string id = 1;
  string title = 2;
  string summary = 3;
  int32 priority = 4;
  string status = 5;
}

message GetProjectByIdRequest { string id = 1; }

message ListProjectsRequest {
  map<string, string> filters = 1; 
}

message BulkCreateProjectResponse {
  int32 projects_created = 1;
  repeated Project projects = 2;
}

service ProjectService {
  rpc CreateProject (CreateProjectRequest) returns (Project);                          
  rpc UpdateProject (UpdateProjectRequest) returns (Project);                          
  rpc DeleteProject (GetProjectByIdRequest) returns (google.protobuf.Empty);
  rpc GetProjectById (GetProjectByIdRequest) returns (Project);
  rpc ListProjects (ListProjectsRequest) returns (stream Project);  
  rpc BulkCreateProjects (stream CreateProjectRequest) returns (BulkCreateProjectResponse);
}
